### PR TITLE
Fix the SIL incoming phi value API, verify critical edges, fix SIL passes that break verification.

### PR DESCRIPTION
**Pull request by atrick**
Simplify SILPHIArgument::getIncomingValue.

The client of this interface naturally expects to get back the
incoming phi value. Ignoring dominance and SIL ownership, the incoming
phi value and the block argument should be substitutable.

This method was actually returning the incoming operand for
checked_cast and switch_enum terminators, which is deeply misleading
and has been the source of bugs.

If the client wants to peek though casts, and enums, it should do so
explicitly. getSingleTerminatorOperand\[s]() will do just that.

### CODE REVIEW

**Comment by atrick** [comment]
Only the last commit is part of this PR. The first 3 commits should be on master soon.

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
@swift-ci test.

**Comment by atrick** [comment]
Running additional verification and fixing optimizer passes before committing this.

**Comment by atrick** [comment]
@swift-ci test source compatibility.

**Comment by atrick** [comment]
@swift-ci smoke benchmark staging.

**Comment by atrick** [comment]
@swift-ci benchmark.

**Comment by atrick** [comment]
@swift-ci smoke benchmark.

**Comment by atrick** [comment]
@gottesmm is already planning to fix SILPHIArgument so that it actually means "phi argument". (It would be nice to correct the spelling at some point, since Phi is not an acronym). Anyway, this PR feeds into that but I intentionally avoided changing type names here.

@eeckstein @aschwaighofer may want to have a look.

**Comment by atrick** [comment]
@swift-ci smoke test.

**Comment by atrick** [comment]
@swift-ci smoke test.

**Review by eeckstein — COMMENTED** [review]
mainly coding style comments, otherwise lgtm

File: include/swift/SIL/SILArgument.h [hunk 1]
```diff
@@ -77,41 +77,50 @@ class SILArgument : public ValueBase {
...
-  /// payload argument is the enum itself (the operand of the switch_enum).
-  SILValue getSingleIncomingValue() const;
+  /// Returns true if we were able to find a single terminator operand value for
+  /// each predecessor of this arguments basic block. The found values are
+  /// stored in OutArray.
+  ///
+  /// Note: this peeks through any projections or cast implied by the
+  /// terminator. e.g. the incoming value for a switch_enum payload argument is
+  /// the enum itself (the operand of the switch_enum).
+  bool getSingleTerminatorOperands(llvm::SmallVectorImpl<SILValue> &OutArray);
```
[the remark this answers is not in the document]
Comment (atrick) [reply include/swift/SIL/SILArgument.h]: You may have been reading the old comment. The new comment explains that this API peeks through "single operand terminators". I don't particularly like this API, but didn't want to rewrite the passes that use the API in this PR.

File: include/swift/SIL/SILArgument.h [hunk 2]
```diff
@@ -77,41 +77,50 @@ class SILArgument : public ValueBase {
...
-
-  /// Returns true if we were able to find incoming values for each predecessor
-  /// of this arguments basic block. The found values are stored in OutArray.
-  /// Note that for some predecessor terminators the incoming value is not
-  /// exactly the argument value. E.g. the incoming value for a switch_enum
-  /// payload argument is the enum itself (the operand of the switch_enum).
-  bool getIncomingValues(
+  /// Return true if this block argument is actually a phi argument as
+  /// opposed to a cast or projection.
+  bool isPhiArgument();
```
[the remark this answers is not in the document]
Comment (atrick) [reply include/swift/SIL/SILArgument.h]: I wholeheartedly agree that `SILPHIArgument` should have been called `SILBlockArgument` and have always argued that. However, @gottesmm already plans to change the meaning of `SILPhiArgument` (notice my spelling change) and introduce an additional `SILTerminatorResultArgument` (or something like that). At that point, I'm not sure if we still need a separate `SILBlockArgument`.

**Comment by atrick** [comment]
@swift-ci smoke test.

**Pull request merged**